### PR TITLE
Fyle 28mekc8 - Revamp of policy and critical policy violation UI

### PR DESCRIPTION
**Pull request by saitejasFyle**
Policy violations will look like this:

Screenshot 2022-04-26 at 7 43 03 PM

Critical policy violations will look like this

Screenshot 2022-04-27 at 2 03 09 PM

### CODE REVIEW

File: src/app/shared/components/fy-critical-policy-violation/fy-critical-policy-violation-message/fy-critical-policy-violation-message.component.scss [hunk 1]
```diff
@@ -0,0 +1,9 @@
+.critical-policy-violation-message {
+  &--container {
+    padding: 16px;
+    border-radius: 3px;
+    background: #fef6f6;
+    border: 1px solid #fbe8e9;
```
Comment (Julias0) [inline src/app/shared/components/fy-critical-policy-violation/fy-critical-policy-violation-message/fy-critical-policy-violation-message.component.scss:6]: use colors from variables inside colors.scss file

File: src/app/shared/components/fy-critical-policy-violation/fy-critical-policy-violation.component.html [hunk 2]
```diff
@@ -1,61 +1,75 @@
-<div class="critical-policy-violation">
-  <ion-header>
-    <div class="critical-policy-violation--header text-center">Critical Policy Violation</div>
-    <div class="critical-policy-violation--info text-center">
-      This expense violates the below policies/guidelines of your company. Kindly resolve the issues to be able to
-      submit this expense.
-    </div>
-  </ion-header>
+<!-- <div class="critical-policy-violation"> -->
```
Comment (Julias0) [inline src/app/shared/components/fy-critical-policy-violation/fy-critical-policy-violation.component.html:1]: avoid commented code

File: src/app/shared/components/fy-critical-policy-violation/fy-critical-policy-violation.component.html [hunk 3]
```diff
@@ -1,61 +1,75 @@
...
-        <ng-container *ngFor="let criticalViolationMessage of criticalViolationMessages">
-          <li class="critical-policy-violation--list__elements">
-            {{ criticalViolationMessage }}
-          </li>
-        </ng-container>
-      </ul>
-    </div>
+    </ng-container>
+
+    <!-- <div class="critical-policy-violation--body">
```
Comment (Julias0) [inline src/app/shared/components/fy-critical-policy-violation/fy-critical-policy-violation.component.html:22]: please remove dead code

File: src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  isExpenseCapped: boolean;
+
+  additionalApprovalString: string;
+
+  cappedAmountString: string;
+
+  constructor(private modalController: ModalController) {}
+
+  ngOnInit() {
+    // this.policyActionDescription =
```
Comment (Julias0) [inline src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts:32]: remove dead code

File: src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      }
+      if (this.policyActionDescription.toLowerCase().includes('primary approver will be skipped')) {
+        this.isPrimaryApproverSkipped = true;
+      }
+      if (this.policyActionDescription.toLowerCase().includes('expense will need additional approval from')) {
+        this.needAdditionalApproval = true;
+        const emails = this.policyActionDescription.match(/([a-zA-Z0-9._-]+@[a-zA-Z0-9._-]+\.[a-zA-Z0-9_-]+)/gi);
+        if (emails?.length > 0) {
+          this.additionalApprovalString = 'Expense will need additional approval from ';
+          let index = 0;
```
Comment (Julias0) [inline src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts:47]: why is this required?

File: src/app/shared/components/fy-policy-violation/policy-violation-message/policy-violation-message.component.html [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<div class="policy-violation-message--container">
+  <div class="policy-violation-message--content" [innerHTML]="message"></div>
```
Comment (Julias0) [inline src/app/shared/components/fy-policy-violation/policy-violation-message/policy-violation-message.component.html:2]: why use innerHTML? Why not {{message}}

Comment (saitejasFyle) [reply src/app/shared/components/fy-policy-violation/policy-violation-message/policy-violation-message.component.html:2]: We have a case where the approver email IDs has to be bold. For that I'm constructing an HTML string. To populate the HTML strings as it is, I've used this since it could be used for normal strings as well

File: src/app/shared/components/fy-critical-policy-violation/fy-critical-policy-violation-message/fy-critical-policy-violation-message.component.scss (hunk 1 above)
Comment (saitejasFyle) [reply src/app/shared/components/fy-critical-policy-violation/fy-critical-policy-violation-message/fy-critical-policy-violation-message.component.scss:6]: Done

File: src/app/shared/components/fy-critical-policy-violation/fy-critical-policy-violation.component.html (hunk 2 above)
Comment (saitejasFyle) [reply src/app/shared/components/fy-critical-policy-violation/fy-critical-policy-violation.component.html:1]: Done

File: src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts (hunk 4 above)
Comment (saitejasFyle) [reply src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts:32]: Done

File: src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+  cappedAmountString: string;
+
+  constructor(private modalController: ModalController) {}
+
+  ngOnInit() {
+    // this.policyActionDescription =
+    //   'The policy violation will trigger the following action(s): expense will be flagged for verification and approval, primary approver will be skipped, expense will need additional approval from dimple.kh@fyle.in, aiyush.dhar@fyle.in, expense will be capped to USD 100.3';
+
+    if (this.policyActionDescription) {
```
[the remark this answers is not in the document]
Comment (saitejasFyle) [reply src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts:35]: Done

File: src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts (hunk 5 above)
Comment (saitejasFyle) [reply src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts]: Removed it and used the forEach arg itself

File: src/app/core/services/policy.service.ts [hunk 8]
```diff
@@ -37,4 +37,36 @@ export class PolicyService {
     };
     return this.policyApiService.get('/policy/violating_transactions', { params });
   }
+
+  isExpenseFlagged(policyActionDescription): boolean {
+    if (policyActionDescription.toLowerCase().includes('expense will be flagged')) {
```
Comment (Julias0) [inline src/app/core/services/policy.service.ts:42]: return policyActionDescription.toLowerCase().includes('expense will be flagged');

same for the rest of the methods here

File: src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+  ngOnInit() {
+    if (this.policyActionDescription) {
+      this.isExpenseFlagged = this.policyService.isExpenseFlagged(this.policyActionDescription);
+      this.isPrimaryApproverSkipped = this.policyService.isPrimaryApproverSkipped(this.policyActionDescription);
+      this.needAdditionalApproval = this.policyService.needAdditionalApproval(this.policyActionDescription);
+      this.isExpenseCapped = this.policyService.isExpenseCapped(this.policyActionDescription);
+
+      if (this.needAdditionalApproval) {
+        const emails = this.policyActionDescription.match(/([a-zA-Z0-9._-]+@[a-zA-Z0-9._-]+\.[a-zA-Z0-9_-]+)/gi);
```
Comment (Julias0) [inline src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts:40]: const emailRegex = /([a-zA-Z0-9._-]+@[a-zA-Z0-9._-]+\.[a-zA-Z0-9_-]+)/gi;
const emails = this.policyActionDescription.match(emailRegex);

File: src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+            this.additionalApprovalString += '<b>' + email + '</b>';
+            if (index < emails.length - 1) {
+              this.additionalApprovalString += ', ';
+            }
+          });
+        }
+      }
+
+      if (this.isExpenseCapped) {
+        const cappedAmount = this.policyActionDescription.match(/capped to ([a-zA-Z]{1,3} \d+)/i)[1];
```
Comment (Julias0) [inline src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts:53]: const amountRegex = ([a-zA-Z]{1,3} \d+)/i;
Same as above. 
You can choose to put the regex inside utilityService and fetch it from there using a method

**Review by Julias0 — APPROVED** [review]

File: src/app/core/services/policy.service.ts (hunk 8 above)
Comment (saitejasFyle) [reply src/app/core/services/policy.service.ts:42]: Done

File: src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts (hunk 10 above)
Comment (saitejasFyle) [reply src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts:53]: Done

File: src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts (hunk 9 above)
Comment (saitejasFyle) [reply src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts:40]: Done

File: src/app/core/services/policy.service.ts [hunk 11]
```diff
@@ -37,4 +37,20 @@ export class PolicyService {
     };
     return this.policyApiService.get('/policy/violating_transactions', { params });
   }
+
+  isExpenseFlagged(policyActionDescription): boolean {
```
Comment (Julias0) [inline src/app/core/services/policy.service.ts:41]: policyActionDescription: string
Add data type to the input arguments

File: src/app/core/services/utility.service.ts [hunk 12]
```diff
@@ -111,6 +111,14 @@ export class UtilityService {
     });
   }
 
+  getEmailsFromString(text: String) {
```
Comment (Julias0) [inline src/app/core/services/utility.service.ts:114]: add return types to these methods

File: src/app/shared/components/fy-policy-violation/policy-violation-message/policy-violation-message.component.scss [hunk 13]
```diff
@@ -0,0 +1,9 @@
+.policy-violation-message {
+  &--container {
+    padding: 16px;
+    border-radius: 3px;
+    background: #fffcf5;
+    border: 1px solid #ffefb7;
```
Comment (Julias0) [inline src/app/shared/components/fy-policy-violation/policy-violation-message/policy-violation-message.component.scss:6]: colors should be in variables

Comment (Julias0) [reply src/app/shared/components/fy-policy-violation/policy-violation-message/policy-violation-message.component.scss:6]: and you should get them from colors.scss

File: src/app/shared/components/fy-policy-violation/policy-violation-action/policy-violation-action.component.scss [hunk 14]
```diff
@@ -0,0 +1,17 @@
+.policy-violation-action {
+  &--container {
+    padding: 16px;
+    border: 1px solid #eff5fc;
```
Comment (Julias0) [inline src/app/shared/components/fy-policy-violation/policy-violation-action/policy-violation-action.component.scss:4]: colors should be from colors.scss

File: src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+      this.isExpenseFlagged = this.policyService.isExpenseFlagged(this.policyActionDescription);
+      this.isPrimaryApproverSkipped = this.policyService.isPrimaryApproverSkipped(this.policyActionDescription);
+      this.needAdditionalApproval = this.policyService.needAdditionalApproval(this.policyActionDescription);
+      this.isExpenseCapped = this.policyService.isExpenseCapped(this.policyActionDescription);
+
+      if (this.needAdditionalApproval) {
+        const emails = this.utilityService.getEmailsFromString(this.policyActionDescription);
+        if (emails?.length > 0) {
+          this.additionalApprovalString = 'Expense will need additional approval from ';
+          emails.forEach((email, index) => {
```
Comment (Julias0) [inline src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts:48]: this.additionalApprovalString = this.getApprovalString(emails);

Comment (saitejasFyle) [reply src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts:48]: Done

File: src/app/shared/components/fy-policy-violation/policy-violation-action/policy-violation-action.component.scss (hunk 14 above)
Comment (saitejasFyle) [reply src/app/shared/components/fy-policy-violation/policy-violation-action/policy-violation-action.component.scss:4]: Done

File: src/app/shared/components/fy-policy-violation/policy-violation-message/policy-violation-message.component.scss (hunk 13 above)
Comment (saitejasFyle) [reply src/app/shared/components/fy-policy-violation/policy-violation-message/policy-violation-message.component.scss:6]: Done

File: src/app/core/services/policy.service.ts (hunk 11 above)
Comment (saitejasFyle) [reply src/app/core/services/policy.service.ts:41]: Done

File: src/app/core/services/utility.service.ts (hunk 12 above)
Comment (saitejasFyle) [reply src/app/core/services/utility.service.ts:114]: Done

File: src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+              'Expense will be capped to ' +
+              getCurrencySymbol(cappedAmountSplit[0], 'wide', 'en') +
+              cappedAmountSplit[1];
+          }
+        }
+      }
+    }
+  }
+
+  getApprovalString(emails) {
```
Comment (Julias0) [inline src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts:67]: `getApprovalString(emails: string[]): string {`

you can also consider moving this into a service

**Review by Julias0 — APPROVED** [review]

File: src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts (hunk 16 above)
Comment (saitejasFyle) [reply src/app/shared/components/fy-policy-violation/fy-policy-violation.component.ts:67]: Done
